[v10] Add checkbox "all" option

frankieroberto opened the pull request:
This updates the Checkbox component with a new "all" option using using `inclusive: true`.

The new "all" checkbox will select all the others when checked.

The existing "none" option deselects all the others when checked.

colinrotherham committed:
Add checkbox "all" behaviour

frankieroberto committed:
Update behaviour for inclusive checkbox

This updates the behaviour for inclusive ("select all") checkboxes, so that if you individually check all checkboxes except for "select all", then the "select all" checkbox gets automatically check when you check the last checkbox.

This avoids being in a state where "select all" is the only checkbox unchecked, and when you check it then nothing happens.

Co-authored-by: Colin Rotherham <work@colinr.com>

frankieroberto commented:
We tested the interface in https://github.com/nhsuk/nhsuk-frontend/pull/1707#issuecomment-3632066144 in our last round of user research, with staff who use Record a vaccination across multiple pharmacies. We didn’t find any issues with the interaction.

In our current prototype I’m always including the select-all option so long as there’s 2+ options (if there’s only a single option they skip the screen entirely).

It might be that it only makes sense to include the select-all option if there are 5+ or some other number of options though.

frankieroberto committed:
Update tests

colinrotherham committed:
Prefer "None of the above" to "None of these"

colinrotherham committed:
Add checkboxes `behaviour` and `behaviourGroup` macro options

colinrotherham committed:
Remove checkboxes `inclusive` and `inclusiveGroup` macro options

colinrotherham committed:
Deduplicate checkboxes code

colinrotherham committed:
Update review app examples

frankieroberto commented:
@colinrotherham I think it makes sense to not show the 'select all' checkbox where javascript is unavailable, and treat it as progressive enhancement? Otherwise, we'd have to infer that if someone ticked 'select all' but none (or only some) of the other checkboxes, which feels slightly risky?

This is different from the 'none' checkbox, as that still has a benefit with no javascript - and we have some guidance suggesting showing an error if they selected both another checkbox _and_ 'none':

> If JavaScript is unavailable, and a user selects both the "none" checkbox and another checkbox, display an error message.

colinrotherham commented:
Thanks @frankieroberto, what did you think about the implementation?

> See commits https://github.com/nhsuk/nhsuk-frontend/commit/895e722cd9481ba8b45c449277852a10b70f549d and https://github.com/nhsuk/nhsuk-frontend/commit/77d95e4997ae359619bbed48d16f66c624b36ab5 – thoughts?

frankieroberto commented:
@colinrotherham I'm happy with the implementation. Adding `behaviour: 'inclusive'` to the divider to also hide that when javascript is disabled makes sense to me, and I can’t think of any other clean way to infer when dividers should be hidden, without a ton of extra logic.

colinrotherham commented:
@frankieroberto Brilliant. I'll need to rebase with `support/10.x` instead of `main`

Will pick it up when I get chance, thank you 🙏

frankieroberto committed:
Add changelog entry

Co-authored-by: Colin Rotherham <work@colinr.com>

frankieroberto commented:
@colinrotherham this is currently a breaking change, so should target `main` instead? 

It's breaking as it removes the existing `exclusive: true`, renaming it to `behaviour: "exclusive"`.

We could make it backwars compatible, and support both `exclusive: true` AND `behaviour: "exclusive"` for "none" options, but perhaps only `behaviour: "inclusive"` for the "all" option? But then that does pose the question of what happens if you use `exclusive: true` and `behaviour: "inclusive"` on the same item... 😬 

If v11 is not far away, it might be simpler to wait until then? But if it's a way off, we could put the effort in to make it backwards-compatible to enable a sooner release?

colinrotherham commented:
Wasn't it backwards compatible until https://github.com/nhsuk/nhsuk-frontend/pull/1707/changes/0709320caf14b872132a210723cfe8178fe01ddf when `exclusive` and `exclusiveGroup` were removed?

Only because this PR has the v10.6.0 milestone

frankieroberto commented:
@colinrotherham at the moment this PR drops support for `exclusive: true`, by the looks of it, so that's not backwards-compatible, but the JavaScript is: https://github.com/nhsuk/nhsuk-frontend/pull/1707/changes#diff-f3029a8bee8b96cb9d973cff3a8a09ae04dd4b7059d231834c1f00487a773654R312

frankieroberto committed:
Add deprecated option

frankieroberto commented:
ok I think we've decided to make this a non-breaking change now, so the older params are still supported but deprecated.

frankieroberto committed:
Add tests and fix bug

Fixes bug: When an inclusive checkbox ("Select all") was unchecked with checkboxes that have unique names and use `inclusiveGroup`, the other checkboxes were not being unchecked because `handleClick` called `unCheckAllInputsExcept()` which uses `checkboxExclusiveGroup` instead of `checkboxInclusiveGroup`.

Co-authored-by: Colin Rotherham <work@colinr.com>

frankieroberto committed:
Make the Nunjucks backwards compatible

Co-authored-by: Colin Rotherham <work@colinr.com>

colinrotherham commented:
@frankieroberto All green now and ready for v10.6.0

I'll do some more testing of this before we merge it in

Thanks 🙏

frankieroberto commented:
@colinrotherham thank you! I’ll draft some guidance.

frankieroberto commented:
Added draft guidance in https://github.com/nhsuk/nhsuk-service-manual/pull/2583

colinrotherham commented:
I've had a little go at updating the examples in the changelog entry now the guidance is coming along

Review app is enabled again too

colinrotherham committed:
Hide checkbox inclusive behaviour items (and dividers) when not supported

colinrotherham commented:
@frankieroberto We previously added `nhsuk-frontend-supported-only` to this PR

It's now been renamed to `nhsuk-u-hidden-not-supported` and split out into https://github.com/nhsuk/nhsuk-frontend/pull/1972

colinrotherham committed:
Update changelog entry

colinrotherham committed:
Prefer shorter "none" and "all" names

colinrotherham commented:
I've renamed our example names to use "none" and "all" (to match the guidance) in https://github.com/nhsuk/nhsuk-frontend/pull/1707/commits/d4d0a1a2d8cddcce518a8e54c46eb3b020d6f177

@frankieroberto Shall we change the colours example to a vaccines one?

colinrotherham reviewed (approved):
Brilliant work! 🙌

Just that question about maybe a more NHS-specific example @frankieroberto

colinrotherham commented:
Also flagging that we don't have Puppeteer tests for "all" if you think we should

colinrotherham committed:
Fix tests to work with more than 3 checkboxes

colinrotherham committed:
Update reference images

anandamaryon1 committed:
Swap colours for vaccines in example

colinrotherham commented on packages/nhsuk-frontend/src/nhsuk/components/checkboxes/checkboxes.jsdom.test.mjs:
```diff
@@ -269,7 +269,7 @@ describe('Checkboxes', () => {
       expect($inputBehaviour).not.toBeChecked()
 
-      // Check "None of the above"
+      // Check "None"
```
Should this be `Check "I do not want to be contacted"`?

Otherwise maybe `Check "none" option` to keep it generic

frankieroberto replied:
generic sounds good to me. Changed in 930436c465d83b8553c36caacb98755351437065

colinrotherham commented on packages/nhsuk-frontend/src/nhsuk/components/checkboxes/checkboxes.mjs:
```diff
@@ -104,7 +104,7 @@ export class Checkboxes extends ConfigurableComponent {
    *
    * Find any other checkbox inputs with the checkbox group value, and uncheck them.
-   * This is useful when a “None of the above" checkbox is checked.
+   * This is useful when a “None of these" checkbox is checked.
```
I think we used to say "None of these" before I changed it to "None of the above"

Shall we refer to the "none" option here too?

frankieroberto replied:
I think "None of these" is ok here? 🤷  Maybe it could be:

```suggestion
   * This is useful when a "None of these" type of checkbox is checked.
```

to make it clear that you don’t have to use that exact wording?

frankieroberto replied:
Or this?

```suggestion
   * This is useful when a "none" checkbox option is checked.
```

colinrotherham replied:
Last one but lowercase "none" like in the guidance?

frankieroberto replied:
done

colinrotherham commented on packages/nhsuk-frontend/src/nhsuk/components/checkboxes/checkboxes.mjs:
```diff
@@ -104,80 +104,180 @@ export class Checkboxes extends ConfigurableComponent {
+   *
+   * Find any checkbox inputs with the same checkbox group value and behaviour
+   * and uncheck them. This helps prevent checking both a regular checkbox and
+   * "None of these" at the same time
```
Got one more "None of these here" 🙈

colinrotherham commented on packages/nhsuk-frontend/src/nhsuk/components/checkboxes/checkboxes.mjs:
```diff
@@ -104,80 +104,180 @@ export class Checkboxes extends ConfigurableComponent {
+   * Check other checkboxes
+   *
+   * Find any other checkbox inputs with the checkbox group value, and check them.
+   * This is useful for when a "Select all" checkbox is checked.
```
```suggestion
   * This is useful for when an "all" checkbox option is checked.
```

colinrotherham commented on packages/nhsuk-frontend/src/nhsuk/components/checkboxes/checkboxes.mjs:
```diff
@@ -104,80 +104,180 @@ export class Checkboxes extends ConfigurableComponent {
+   * Check inputs
+   *
+   * Find any checkbox inputs with the same checkbox group value and behaviour
+   * and check them. This helps prevent "Select all" being left unchecked when
```
```suggestion
   * and check them. This helps prevent "all" options being left unchecked when
```

frankieroberto committed:
Update reference images

frankieroberto commented:
@colinrotherham @anandamaryon1 all issues resolved. Ok to merge this?

colinrotherham commented on CHANGELOG.md:
```diff
@@ -3993,7 +4094,7 @@ After:
 ### :new: **New features**
 
-- Add "None of these" JavaScript behaviour to checkboxes - allowing all checkboxes in a group to be automatically unchecked when the "None of the above" option is checked ([PR 756](https://github.com/nhsuk/nhsuk-frontend/pull/756))
+- Add "None of these" JavaScript behaviour to checkboxes - allowing all checkboxes in a group to be automatically unchecked when the "None of these" option is checked ([PR 756](https://github.com/nhsuk/nhsuk-frontend/pull/756))
```
We should leave old changelog entries as they were, I'll edit this

colinrotherham commented on CHANGELOG.md:
```diff
@@ -1812,9 +1913,9 @@ The file structure for the stylesheets has changed. You will have to make differ
 For precompiled CSS files, note the following path changes:
 
-- copy or serve `node_modules/nhsuk-frontend/dist/nhsuk/nhsuk-frontend.min.css`  
+- copy or serve `node_modules/nhsuk-frontend/dist/nhsuk/nhsuk-frontend.min.css`
```
We've lost the double spaces (to preserve line breaks) at the end of a few of these lines

I'll sort it

frankieroberto committed:
Reword "None of the above"

This is to avoid using the spatial language, as it can be vague, and might be confusing for screenreader users.

The "none" checkbox could also technically be in any position in the list (although conventionally will be last).

colinrotherham committed:
Add test for deprecated method

colinrotherham commented:
@frankieroberto I've updated the macro option descriptions to use "all" and "none" in https://github.com/nhsuk/nhsuk-frontend/pull/1707/commits/2fb2291408cd877489adbc77b91da9552da0c3ce  and added a test for the deprecated method in https://github.com/nhsuk/nhsuk-frontend/pull/1707/commits/416836be0e7e715706df3530716045acd1260eb4. Think we're good to go now
